spanner/spanner_snippets/spanner/spanner_opentelemetry_usage.go

```diff
@@ -0,0 +1,130 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spanner
+
+// [START spanner_opentelemetry_usage]
+// Ensure that your Go runtime version is supported by the OpenTelemetry-Go compatibility policy before enabling OpenTelemetry instrumentation.
+// Refer to compatibility here https://github.com/googleapis/google-cloud-go/blob/main/debug.md#opentelemetry
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+
+	"cloud.google.com/go/spanner"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+	"google.golang.org/api/iterator"
+)
+
+func enableOpenTelemetryMetricsAndTraces(w io.Writer, db string) error {
+	// db = `projects/<project>/instances/<instance-id>/database/<database-id>`
+	ctx := context.Background()
+
+	// Create a new resource to uniquely identify the application
+	res, err := newResource()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Enable OpenTelemetry traces by setting environment variable GOOGLE_API_GO_EXPERIMENTAL_TELEMETRY_PLATFORM_TRACING=opentelemetry at the start of the application
+	// Enable OpenTelemetry metrics before injecting meter provider.
+	spanner.EnableOpenTelemetryMetrics()
+
+	// Create a new tracer provider
+	tracerProvider, err := getOtlpTracerProvider(ctx, res)
+	defer tracerProvider.ForceFlush(ctx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// Register tracer provider as global
+	otel.SetTracerProvider(tracerProvider)
+
+	// Create a new meter provider
+	meterProvider := getOtlpMeterProvider(ctx, res)
+	defer meterProvider.ForceFlush(ctx)
+
+	// Inject meter provider locally via ClientConfig when creating a spanner client.
+	client, err := spanner.NewClientWithConfig(ctx, db, spanner.ClientConfig{OpenTelemetryMeterProvider: meterProvider})
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	stmt := spanner.Statement{SQL: `SELECT SingerId, AlbumId, AlbumTitle FROM Albums`}
+	iter := client.Single().Query(ctx, stmt)
+	defer iter.Stop()
+	for {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		var singerID, albumID int64
+		var albumTitle string
+		if err := row.Columns(&singerID, &albumID, &albumTitle); err != nil {
+			return err
+		}
+		fmt.Fprintf(w, "%d %d %s\n", singerID, albumID, albumTitle)
+	}
+}
+
+func getOtlpMeterProvider(ctx context.Context, res *resource.Resource) *metric.MeterProvider {
+	otlpExporter, err := otlpmetricgrpc.New(ctx, otlpmetricgrpc.WithInsecure())
```

Comment:
does this require `otlpmetricgrpc.WithInsecure()`? This does not seem like the right path to recommend, and I have not noticed this in other otel samples.

Reply:
The reason for adding this is because OTLP collector is non-TLS enabled and OTLP exporter is TLS enabled. This mismatch does not make the above sample directly work and need some authentication setup by following OpenTelemetry documentation. Adding this `insecure` will make the exporter non-TLS and customers can directly run the sample without issues.
I have got consensus from OpenTelemetry team on this.

PLMK if this is not the right way. I can remove it if you feel so.

Reply:
I have removed this from the sample.